File: example/avx/count.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"io"
	"bytes"
	"time"
)

func lineCounter(r io.Reader) (int, error) {
	buf := make([]byte, 2*1024*1024)
	count := 0
	lineSep := []byte{'\n'}
	for {
		c, err := r.Read(buf)
		count += bytes.Count(buf[:c], lineSep)
		switch {
		case err == io.EOF:
			return count, nil
		case err != nil:
			return count, err
		}
	}
}

func count_lines_in_file (filename string) (int, error) {
	file, err := os.Open(filename)
	if err != nil {
		return 0, err
	}
	lineCount, err := lineCounter(bufio.NewReader(file))
	if err != nil {
		return 0, err
	}
	file.Close()
	return lineCount, nil
}

func main() {
	var (
		filename string
	)
	filename = "/dev/shm/test.log"
	if (len(os.Args) > 1) {
		filename = os.Args[1]
	}
	expected, err := count_lines_in_file(filename)
	if err != nil {
		log.Fatalf("Error: %v", err)
	}
	fmt.Printf("%d\n", expected)
	for true {
		start := time.Now()
		lines, err := count_lines_in_file(filename)
		elapsed := time.Since(start)
		if err != nil {
			log.Fatalf("Error: %v", err)
		}
		if lines != expected {
			log.Fatalf("line count is incorrect")
		}
		fmt.Printf("time %d ns\n", elapsed.Nanoseconds())
	}
}

File: example/avx/count.js
const { memcount } = lo.load('memcount')
const { core, assert } = lo
const { count_avx } = memcount
const { 
  close, open, read2, mmap, madvise, 
  O_RDONLY, MAP_PRIVATE, MAP_ANONYMOUS, MADV_HUGEPAGE, PROT_READ, PROT_WRITE
} = core
const memory_flags = MAP_ANONYMOUS | MAP_PRIVATE

function create_buffer (size, prot = PROT_READ | PROT_WRITE) {
  const ptr = assert(mmap(0, size, prot, memory_flags, -1, 0))
  assert(madvise(ptr, size, MADV_HUGEPAGE) === 0)
  return ptr
}

function count_chars (file_name) {
  let total = 0
  let bytes = 0
  const fd = open(file_name, O_RDONLY)
  assert(fd > 2)
  while ((bytes = read2(fd, buf_ptr, len)) > 0) {
    total += count_avx(buf_ptr, 10, bytes)
  }
  close(fd)
  return total
}

const len = 2 * 1024 * 1024
const buf_ptr = create_buffer(len)
const file_name = lo.args[2] || '/dev/shm/test.log'
const expected = count_chars(file_name)
console.log(expected)

while (1) {
  const start = lo.hrtime()
  assert(count_chars(file_name) === expected)
  console.log(`time ${lo.hrtime() - start} ns`)
}
